ethereum/blockwatch/client.go

```diff
@@ -54,6 +61,8 @@ type GetBlockByNumberResponse struct {
 // HeaderByNumber fetches a block header by its number. If no `number` is supplied, it will return the latest
 // block header. If no block exists with this number it will return a `ethereum.NotFound` error.
 func (rc *RpcClient) HeaderByNumber(number *big.Int) (*miniheader.MiniHeader, error) {
+	rc.rateLimiter.Wait(context.Background())
```

Comment:
Looking at it again, this is actually feasible. In order to satisfy the client passed to contract wrappers, I only need to satisfy:

```
// ContractCaller defines the methods needed to allow operating with contract on a read
// only basis.
type ContractCaller interface {
	// CodeAt returns the code of the given account. This is needed to differentiate
	// between contract internal errors and the local chain being out of sync.
	CodeAt(ctx context.Context, contract common.Address, blockNumber *big.Int) ([]byte, error)
	// ContractCall executes an Ethereum contract call with the specified data as the
	// input.
	CallContract(ctx context.Context, call ethereum.CallMsg, blockNumber *big.Int) ([]byte, error)
}
```

Since we only make `call`s. I've got ahead and made a `EthRPCClient` that we now use for _all_ Ethereum RPC calls and that takes care of rate limiting under-the-hood 🎊  Much cleaner.